### PR TITLE
🗺️  662 Filter Programs by Region

### PR DESCRIPTION
**Pull request by demariadaniel**
**Description of changes**

Filters programs by requested regions

**Type of Change**

- [ ] Bug
- [ ] New Feature

### CODE REVIEW

**Comment by daniel-cy-lu** [comment]
@joneubank @demariadaniel I have ticket #690 . Where I am to remove the "regions" field in the Program Type. I see Dan is using program.regions as part of his filter and this would be broken if I am to remove regions.

**Comment by demariadaniel** [comment]
> @joneubank @demariadaniel I have ticket #690 . Where I am to remove the "regions" field in the Program Type. I see Dan is using program.regions as part of his filter and this would be broken if I am to remove regions.

Thanks for mentioning, I saw that too, makes things confusing

**Comment by daniel-cy-lu** [comment]
@demariadaniel Jon was at my desk looking at this PR. So we won't be using regions going forward. Instead we will be using dataCenter field. So maybe we do the same thing filtering by dataCenter. I have a ticket #693 that I will be adding dataCenter to the Program type. It had a blocker that just got resolved. I will work on it now, so Dan you can modify your filter.

**Comment by demariadaniel** [comment]
> @demariadaniel @joneubank Jon was at my desk looking at this PR. So we won't be using regions going forward. Instead we will be using dataCenter field. So maybe we do the same thing filtering by dataCenter. I have a ticket #693 that I will be adding dataCenter to the Program type. It had a blocker that just got resolved. I will work on it now, so Dan you can modify your filter.

Thank you! I will revert this to a `draft` PR for now. I will follow your changes!

**Comment by daniel-cy-lu** [comment]
@demariadaniel dataCenter is ready. It's a property in the gql Program type.

**Review by joneubank — CHANGES_REQUESTED** [review]
It would be better to modify the existing `programs` query to add an optional filter by datacenter ID. We've definitely added new queries like this before but you can see the amount of noise that generates.

Can we rebuild it as a filter instead of as a new query?

**Comment by demariadaniel** [comment]
> It would be better to modify the existing `programs` query to add an optional filter by datacenter ID. We've definitely added new queries like this before but you can see the amount of noise that generates.
> 
> Can we rebuild it as a filter instead of as a new query?

Yes, good call. Updated in https://github.com/icgc-argo/platform-api/pull/704/commits/821fde1e991a136dfe5724a86e666348fd41c3c5

File: src/schemas/Program/index.js [hunk 1]
```diff
@@ -352,10 +351,19 @@ const resolvers = {
...
-		programs: async (obj, args, context, info) => {
+		programs: async (obj, args, context) => {
 			const { egoToken } = context;
-			return resolvePrivateProgramList(egoToken);
+			const { dataCenter } = args;
+
+			const programs = await resolvePrivateProgramList(egoToken);
+
+			const filteredPrograms = programs.filter(
+				(program) => !dataCenter || program.dataCenter?.shortName === dataCenter,
```
Comment (ciaranschutte) [inline src/schemas/Program/index.js:361]: I think you can avoid the filter iterations completely if there's no dataCenter arg?

Comment (demariadaniel) [reply src/schemas/Program/index.js]: Yes! Updated

File: src/schemas/Program/index.js [hunk 2]
```diff
@@ -352,10 +351,19 @@ const resolvers = {
...
+		programs: async (obj, args, context) => {
 			const { egoToken } = context;
-			return resolvePrivateProgramList(egoToken);
+			const { dataCenter } = args;
+
+			const programs = await resolvePrivateProgramList(egoToken);
+
+			const filteredPrograms = !dataCenter
+				? programs
+				: programs.filter((program) => program.dataCenter?.shortName === dataCenter);
```
Comment (joneubank) [inline src/schemas/Program/index.js:362]: can you invert the check here? simpler to read with the positive condition:


```ts
			const filteredPrograms = dataCenter
				? programs.filter((program) => program.dataCenter?.shortName === dataCenter)
				: programs;
```

**Review by joneubank — CHANGES_REQUESTED** [review]
Great change, thanks. Minor request here.

File: src/schemas/Program/index.js (hunk 2 above)
Comment (demariadaniel) [reply src/schemas/Program/index.js]: Np sorry about that, updated

**Review by ciaranschutte — APPROVED** [review]
looks good to me

**Comment by ciaranschutte** [comment]
@demariadaniel ah I thought my approval would be enough to merge. looks like you specifically need whoever requested changes approval. @joneubank

**Review by joneubank — APPROVED** [review]

**Pull request merged**